spharpy/spherical.py

```diff
@@ -1,104 +1,599 @@
+            self._recompute_basis_gradient = True
+            self._recompute_inverse = True
+            self._recompute_inverse_gradient = True
+        self._inverse_transform = value
+
+    @property
+    def channel_convention(self):
+        return self._channel_convention
+
+    @channel_convention.setter
+    def channel_convention(self, value):
+        if value not in ["complex", "real"]:
+            raise ValueError("Invalid channel convention, "
+                             "currently only 'acn' "
+                             "and 'fuma' are supported"
+                             )
+        if value != self._channel_convention:
+            self._recompute_basis = True
+            self._recompute_basis_gradient = True
+            self._recompute_inverse = True
+            self._recompute_inverse_gradient = True
+        self._channel_convention = value
+
+    @property
+    def normalization(self):
+        return self._normalization
+
+    @normalization.setter
+    def normalization(self, value):
+        if value not in ["n3d", "maxN", "sn3d"]:
+            raise ValueError(
+                "Invalid normalization, "
+                "currently only 'n3d', 'maxN', 'sn3d' are "
+                "supported"
+            )
+        if value != self._normalization:
+            self._recompute_basis = True
+            self._recompute_basis_gradient = True
+            self._recompute_inverse = True
+            self._recompute_inverse_gradient = True
+        self._normalization = value
+
+    @property
+    def basis(self):
+        if self._basis is None or self._recompute_basis:
+            self._compute_basis()
+            self._recompute_basis = False
+        return self._basis
+
+    @property
+    def basis_gradient_theta(self):
+        if (self._basis_gradient_theta is None or
+                self._recompute_basis_gradient):
+            self._compute_basis_gradient()
+            self._recompute_basis_gradient = False
+        return self._basis_gradient_theta
+
+    @property
+    def basis_gradient_phi(self):
+        if (self._basis_gradient_phi is None or
+                self._recompute_basis_gradient):
+            self._compute_basis_gradient()
+            self._recompute_basis_gradient = False
+        return self._basis_gradient_phi
+
+    # @lru_cache(maxsize=128)
+    def _compute_basis(self):
+        """
+        Compute the basis matrix for the SphericalHarmonics class.
+        """
+        if self.basis_type == "complex":
+            function = spherical_harmonic_basis
+        elif self.basis_type == "real":
+            function = spherical_harmonic_basis_real
+        else:
+            raise ValueError(
+                "Invalid signal type, should be either 'complex' or 'real'"
+            )
+        self._basis = function(
+            self.n_max, self.coordinates,
+            self.normalization, self.channel_convention)
+
+    def compute_basis(self):
```

Comment:
All `compute_` functions should be private and for this example a property `basis` should call the private compute function.